In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Naive Bayes implementation practise \#2
## First some feature engineering

In [2]:
url = 'https://raw.githubusercontent.com/umaimehm/Intro_to_AI/master/machinelearning-Lab4/Lab4/NaiveBayes/Example2/spam_ham_dataset.csv'
dataFrame = pd.read_csv(url)

In [3]:
dataFrame.head()

Unnamed: 0 label                                               text
0         605   ham  Subject: enron methanol ; meter # : 988291\r\n...
1        2349   ham  Subject: hpl nom for january 9 , 2001\r\n( see...
2        3624   ham  Subject: neon retreat\r\nho ho ho , we ' re ar...
3        4685  spam  Subject: photoshop , windows , office . cheap ...
4        2030   ham  Subject: re : indian springs\r\nthis deal is t...

In [4]:
'''
info om .describe()
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.describe.html
'''
dataFrame.groupby('label').describe()

Unnamed: 0                                                         \
           count    mean          std     min      25%     50%      75%   
label                                                                     
ham       3672.0  1835.5  1060.159422     0.0   917.75  1835.5  2753.25   
spam      1499.0  4421.0   432.868340  3672.0  4046.50  4421.0  4795.50   

               
          max  
label          
ham    3671.0  
spam   5170.0

In [5]:
spamDF = dataFrame.copy()

In [6]:
"""
Here 'Lambda' function takes each & every value of the 'label' column & checks if 
value is 'spam'. It will return 1 if spam else 0 (in case of 'ham' return value will be '0')
"""
spamDF['spam'] = spamDF['label'].apply(lambda x: 1 if x=='spam' else 0)

In [7]:
spamDF.head()

Unnamed: 0 label                                               text  spam
0         605   ham  Subject: enron methanol ; meter # : 988291\r\n...     0
1        2349   ham  Subject: hpl nom for january 9 , 2001\r\n( see...     0
2        3624   ham  Subject: neon retreat\r\nho ho ho , we ' re ar...     0
3        4685  spam  Subject: photoshop , windows , office . cheap ...     1
4        2030   ham  Subject: re : indian springs\r\nthis deal is t...     0

In [8]:
X_train, X_test, y_train, y_test = train_test_split(spamDF['text'],spamDF['spam'], test_size=0.30)

In [9]:
X_train

2379    Subject: calpine force majeure\r\nattached is ...
2185    Subject: reduce monthly - payments\r\nthank yo...
2100    Subject: tenaska iv 9 / 01\r\ndarren :\r\nplea...
1826    Subject: hi paliourg software cds 15 and 99 ge...
646     Subject: re : fuel\r\nwhat is the answer ? can...
                              ...                        
1245    Subject: unify production back - eol deals hig...
4649    Subject: viiiicodin no script neeeded\r\nyour ...
2665                                        Subject: \r\n
754     Subject: win a turbocharged jaguar xp in carte...
499                       Subject: no more injections\r\n
Name: text, Length: 3619, dtype: object

### Må gjøre om tekst daten til nummer på et vis, under er en funksjon fra sklearn som kan gjøre dette.

In [18]:
from sklearn.feature_extraction.text import CountVectorizer
v = CountVectorizer()
X_train_count = v.fit_transform(X_train.values)
X_train_count.toarray()[:4]

# Denne koden lager en matrise, disse nummerne representerer unike ord fra data settet
# sklearn CountVectorizer konverterer epost tekst til nummer i en matrise representasjon

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

## Trening, test og evaluering av modellen

In [11]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()
model.fit(X_train_count,y_train)

# Her kjører vi funksjonen "model.fit()" på X_train_count,y_train
# X_train_count inneholder nummer matrisen fra den konverterte teksten.

MultinomialNB()

In [17]:
# evaluerer modellen

emails = [
    'subject: Dear Kenneth, Lets Play football?',
    'subject: 1 billion Nok lottery, for doing nothing, Dont miss this reward!',
]

emails_count = v.transform(emails)
model.predict(emails_count)

# Return = 1 betyr Spam!
# Return = 0 betyr Ham!

array([0, 1])

In [14]:
#Nøyaktighets test

X_test_count = v.transform(X_test)
model.score(X_test_count, y_test)

# (X_test) må transformeres til nummer da modellen ikke klarer å tolke tekst strenger
# Deretter mater vi den til modellen vår

0.9826030927835051

## Denne modellen er 98% nøyaktig
### Konklusjon bør se om det er noe "overfitting"